## Chunking :-

- Different LLM models have different token limit, we can not process millions of tokens at a time

- when we load a large text, will split it into smaller pieces is called as Chunks

- These chunks are then Embedded (numerical representation) and stored in a vector database

- So that relevant chunks can be retrived

- Types of Chunking methods :-
    
    - **CharacterTextSplitter**
    
    - **RecursiveCharacterTextSplitter**
    
    - **TokenTextSplitter**
    
    - **SemanticChunking**
    
    - NLTKTextSplitter
    
    - MarkDownTextSplitter
    
    - SpacyTextSplitter
    
    - PythonTextSplitter
    
    - HTMLTextSplitter

#### Character Text Splitter

- Just split based on characters

In [9]:
from langchain.text_splitter import CharacterTextSplitter

In [13]:
import langchain
langchain.__version__

'0.3.27'

In [15]:
text = """LangChain is an open-source framework designed to facilitate the development of applications that integrate large language models (LLMs) with external data sources and computational logic.
It provides tools, abstractions, and components to help developers build language model-powered applications beyond simple text generation.
LangChain focuses on enabling chains of operations, where outputs from an LLM can trigger additional logic, database queries, API calls, or document retrieval steps.
This chaining approach allows LLMs to perform multi-step reasoning, question-answering, summarization, and decision-making over complex workflows.
"""

In [17]:
Splitter = CharacterTextSplitter()

In [19]:
Splitter.split_text(text)

['LangChain is an open-source framework designed to facilitate the development of applications that integrate large language models (LLMs) with external data sources and computational logic.\nIt provides tools, abstractions, and components to help developers build language model-powered applications beyond simple text generation.\nLangChain focuses on enabling chains of operations, where outputs from an LLM can trigger additional logic, database queries, API calls, or document retrieval steps.\nThis chaining approach allows LLMs to perform multi-step reasoning, question-answering, summarization, and decision-making over complex workflows.']

In [21]:
out = Splitter.split_text(text)
type(out), len(out)

(list, 1)

In [23]:
out

['LangChain is an open-source framework designed to facilitate the development of applications that integrate large language models (LLMs) with external data sources and computational logic.\nIt provides tools, abstractions, and components to help developers build language model-powered applications beyond simple text generation.\nLangChain focuses on enabling chains of operations, where outputs from an LLM can trigger additional logic, database queries, API calls, or document retrieval steps.\nThis chaining approach allows LLMs to perform multi-step reasoning, question-answering, summarization, and decision-making over complex workflows.']

- Above text is splitting by using '\n\n'

In [26]:
Splitter = CharacterTextSplitter(separator='\n')
Splitter.split_text(text)

['LangChain is an open-source framework designed to facilitate the development of applications that integrate large language models (LLMs) with external data sources and computational logic.\nIt provides tools, abstractions, and components to help developers build language model-powered applications beyond simple text generation.\nLangChain focuses on enabling chains of operations, where outputs from an LLM can trigger additional logic, database queries, API calls, or document retrieval steps.\nThis chaining approach allows LLMs to perform multi-step reasoning, question-answering, summarization, and decision-making over complex workflows.']

In [28]:
len(Splitter.split_text(text))

1

In [34]:
Splitter = CharacterTextSplitter(separator=' ')
Splitter.split_text(text)

['LangChain is an open-source framework designed to facilitate the development of applications that integrate large language models (LLMs) with external data sources and computational logic.\nIt provides tools, abstractions, and components to help developers build language model-powered applications beyond simple text generation.\nLangChain focuses on enabling chains of operations, where outputs from an LLM can trigger additional logic, database queries, API calls, or document retrieval steps.\nThis chaining approach allows LLMs to perform multi-step reasoning, question-answering, summarization, and decision-making over complex workflows.']

In [36]:
Splitter = CharacterTextSplitter(separator='')
Splitter.split_text(text)

['LangChain is an open-source framework designed to facilitate the development of applications that integrate large language models (LLMs) with external data sources and computational logic.\nIt provides tools, abstractions, and components to help developers build language model-powered applications beyond simple text generation.\nLangChain focuses on enabling chains of operations, where outputs from an LLM can trigger additional logic, database queries, API calls, or document retrieval steps.\nThis chaining approach allows LLMs to perform multi-step reasoning, question-answering, summarization, and decision-making over complex workflows.']

In [ ]:
CharacterTextSplitter(separator='\n\n')
CharacterTextSplitter(separator='\n')
CharacterTextSplitter(separator=' ')

The above three types expects not only separator 
how much chunk size we need
also we need overlap characters

In [38]:
len('Langchain is an open')

20

- the chunk first preference based on sepeartor only

- below examples for every new line we have more than 20 tokens

- we provided chunk size of 20 tokens, even though because of separater the chunk size is more than 20

- this indicates this method will not take care of sentences and boundaries

In [ ]:
- separater : '\n'

- chunksize : 20

- overlap : 2

In [41]:
len('Langchain is an open')

20

In [49]:
len('Langchain is an open')
len('-source framework de')

20

In [51]:
text

'LangChain is an open-source framework designed to facilitate the development of applications that integrate large language models (LLMs) with external data sources and computational logic.\nIt provides tools, abstractions, and components to help developers build language model-powered applications beyond simple text generation.\nLangChain focuses on enabling chains of operations, where outputs from an LLM can trigger additional logic, database queries, API calls, or document retrieval steps.\nThis chaining approach allows LLMs to perform multi-step reasoning, question-answering, summarization, and decision-making over complex workflows.\n'

In [78]:
Splitter = CharacterTextSplitter(separator='\n',
                                chunk_size = 20, #least preference
                                chunk_overlap = 2)

chunks = Splitter.split_text(text)
chunks

Created a chunk of size 188, which is longer than the specified 20
Created a chunk of size 139, which is longer than the specified 20
Created a chunk of size 165, which is longer than the specified 20


['LangChain is an open-source framework designed to facilitate the development of applications that integrate large language models (LLMs) with external data sources and computational logic.',
 'It provides tools, abstractions, and components to help developers build language model-powered applications beyond simple text generation.',
 'LangChain focuses on enabling chains of operations, where outputs from an LLM can trigger additional logic, database queries, API calls, or document retrieval steps.',
 'This chaining approach allows LLMs to perform multi-step reasoning, question-answering, summarization, and decision-making over complex workflows.']

- If the separator characters < chunk size --> Successful Split

- If the separator characters > chunk size --> Will get a Warnings

    - ex. characters 188 > chunk size 20 --> warnings

In [60]:
len('LangChain is an open-source framework designed to facilitate the development of applications that integrate large language models (LLMs) with external data sources and computational logic.')

188

In [80]:
Splitter = CharacterTextSplitter(separator='.',
                                chunk_size = 20, #least preference
                                chunk_overlap = 2)

chunks = Splitter.split_text("Hello How Are You.I am Good")
chunks

['Hello How Are You', 'I am Good']

In [82]:
Splitter = CharacterTextSplitter(separator='.',
                                chunk_size = 5, #least preference
                                chunk_overlap = 2)

chunks = Splitter.split_text("Hello How Are You.I am Good")
chunks

Created a chunk of size 17, which is longer than the specified 5


['Hello How Are You', 'I am Good']

In [84]:
Splitter = CharacterTextSplitter(separator=' ',
                                chunk_size = 5, #least preference
                                chunk_overlap = 2)

chunks = Splitter.split_text("Hello How Are You.I am Good")
chunks

['Hello', 'How', 'Are', 'You.I', 'am', 'Good']

In [74]:
t1 = 'LangChain is an open-source framework designed to facilitate the development of applications that integrate large language models (LLMs) with external data sources and computational logic.'
from nltk.tokenize import word_tokenize
for t1 in Splitter.split_text(text):
    print(len(word_tokenize(t1)))
    

Created a chunk of size 9, which is longer than the specified 5
Created a chunk of size 11, which is longer than the specified 5
Created a chunk of size 9, which is longer than the specified 5
Created a chunk of size 8, which is longer than the specified 5
Created a chunk of size 10, which is longer than the specified 5
Created a chunk of size 11, which is longer than the specified 5
Created a chunk of size 12, which is longer than the specified 5
Created a chunk of size 9, which is longer than the specified 5
Created a chunk of size 8, which is longer than the specified 5
Created a chunk of size 6, which is longer than the specified 5
Created a chunk of size 6, which is longer than the specified 5
Created a chunk of size 8, which is longer than the specified 5
Created a chunk of size 7, which is longer than the specified 5
Created a chunk of size 13, which is longer than the specified 5
Created a chunk of size 9, which is longer than the specified 5
Created a chunk of size 8, which is

LookupError: 
**********************************************************************
  Resource punkt_tab not found.
  Please use the NLTK Downloader to obtain the resource:

  >>> import nltk
  >>> nltk.download('punkt_tab')
  
  For more information see: https://www.nltk.org/data.html

  Attempted to load tokenizers/punkt_tab/english/

  Searched in:
    - '/Users/pranay/nltk_data'
    - '/opt/anaconda3/nltk_data'
    - '/opt/anaconda3/share/nltk_data'
    - '/opt/anaconda3/lib/nltk_data'
    - '/usr/share/nltk_data'
    - '/usr/local/share/nltk_data'
    - '/usr/lib/nltk_data'
    - '/usr/local/lib/nltk_data'
**********************************************************************


**chunk size means characters**

In [86]:
text = """LangChain is an open-source framework designed to facilitate the development of applications that integrate large language models (LLMs) with external data sources and computational logic.
It provides tools, abstractions, and components to help developers build language model-powered applications beyond simple text generation.
LangChain focuses on enabling chains of operations, where outputs from an LLM can trigger additional logic, database queries, API calls, or document retrieval steps.
This chaining approach allows LLMs to perform multi-step reasoning, question-answering, summarization, and decision-making over complex workflows.
"""

Splitter = CharacterTextSplitter(separator='', # never get as there is no empty
                                chunk_size = 20, #now this is preference
                                chunk_overlap = 2)

chunks = Splitter.split_text(text)
chunks

['LangChain is an open',
 'en-source framework',
 'k designed to facili',
 'litate the developme',
 'ment of applications',
 'ns that integrate la',
 'large language model',
 'els (LLMs) with exte',
 'ternal data sources',
 's and computational',
 'l logic.\nIt provides',
 'es tools, abstractio',
 'ions, and components',
 'ts to help developer',
 'ers build language m',
 'model-powered appli',
 'lications beyond sim',
 'imple text generatio',
 'ion.\nLangChain focus',
 'uses on enabling cha',
 'hains of operations,',
 's, where outputs fro',
 'rom an LLM can trigg',
 'gger additional logi',
 'gic, database querie',
 'ies, API calls, or d',
 'document retrieval',
 'l steps.\nThis chaini',
 'ning approach allows',
 'ws LLMs to perform m',
 'multi-step reasonin',
 'ing, question-answer',
 'ering, summarization',
 'on, and decision-mak',
 'aking over complex w',
 'workflows.']

### why we need overlap

- we seen it is splitting randomly based on seperator and based on chunk

- the sentence is breaking

- model sees one chunk at a time and convert into numerical format

- if we add overlap the end of the chunk is repeated at the start of text

- Context will be continue

- Better retrieval process


### Recursive Text Splitter

- Recursive means hiercachy means step by step process

- It first tries split by paragraphs\n\n).

- If the chunk is still too long it splits by sentence(.)

- If still long then by space or charcter

- Ensure chunks are within the given size but keeps context (using overlap)

- Paragraph ====> Sentence ====> words

In [100]:
from langchain.text_splitter import RecursiveCharacterTextSplitter

text = """LangChain is an open-source framework designed to facilitate the development of applications that integrate large language models (LLMs) with external data sources and computational logic.
It provides tools, abstractions, and components to help developers build language model-powered applications beyond simple text generation.
LangChain focuses on enabling chains of operations, where outputs from an LLM can trigger additional logic, database queries, API calls, or document retrieval steps.
This chaining approach allows LLMs to perform multi-step reasoning, question-answering, summarization, and decision-making over complex workflows.
"""

Splitter = CharacterTextSplitter(separator='', 
                                chunk_size = 50, 
                                chunk_overlap = 2)

chunks = Splitter.split_text(text)
chunks

['LangChain is an open-source framework designed to',
 'o facilitate the development of applications that',
 't integrate large language models (LLMs) with exte',
 'ternal data sources and computational logic.\nIt pr',
 'provides tools, abstractions, and components to he',
 'help developers build language model-powered appli',
 'lications beyond simple text generation.\nLangChain',
 'in focuses on enabling chains of operations, where',
 're outputs from an LLM can trigger additional logi',
 'gic, database queries, API calls, or document retr',
 'trieval steps.\nThis chaining approach allows LLMs',
 's to perform multi-step reasoning, question-answer',
 'ering, summarization, and decision-making over com',
 'omplex workflows.']

In [102]:
len('LangChain is an open-source framework designed to')

49

In [104]:
len('o facilitate the development of applications that')

49

- Paragraph

- Sentence

- Spaces

- Empty Strings

In [107]:
separators = ['\n\n','.',' ','']

text="""Paragraph one: This is the first paragraph. It has two sentences.
Paragraph two: This is the second paragrah which is quite long and will exceed our chunk size limit
for demenostration purpose. It has several sentences to ensure splitting. Here another sentence to make it longer.
Paragraph three: Short."""

In [109]:
text="""Paragraph one: This is the first paragraph. It has two sentences.
Paragraph two: This is the second paragrah which is quite long and will exceed our chunk size limit
for demenostration purpose. It has several sentences to ensure splitting. Here another sentence to make it longer.
Paragraph three: Short."""

# Paragraph level : '\n\n'
len('Paragraph one: This is the first paragraph. It has two sentences.')

65

In [111]:
len('Paragraph two: This is the second paragrah which is quite long and will exceed our chunk size limit for demenostration purpose. It has several sentences to ensure splitting. Here another sentence to make it longer.')

214

In [113]:
len('Paragraph three: Short.')

23

- if will it split at paragraph level we have following chunk size

- 65, 214, 23

- but if we provide chunk size has 50 only

- then paragraph 1 and paragraph 2 are large so further splitting will happend

- that further split based on sentences

#### Sentence level split understanding

In [119]:
len('Paragraph one: This is the first paragraph.')

43

In [121]:
len('It has two sentences.')

21

In [123]:
len('Paragraph one: This is the first paragraph.'),len('It has two sentences.')

(43, 21)

In [125]:
len('Paragraph two: This is the second paragrah which is quite long and will exceed our chunk size limit for demenostration purpose.'), len('It has several sentences to ensure splitting.'), len('Here another sentence to make it longer.')

(127, 45, 40)

In [127]:
len('Paragraph three: Short.')

23

- Now first paragraph does not have any problem, it will divide into two chunks

    - 43, 21 and both are less than 50

- Second one splitted into 3 chunks
    
    - 127, 45 and 40. But first chunk has 127 characters, it is more than 50
    
    - So further split require
    
    - further splitting means space

### Space level split

- logically there are so many spaces will available

- it will concentrate till max char to reach 50

In [139]:
len('''Paragraph two: This is the second paragrah which is quite long and will exceed our chunk size limit for demenostration purpose.
It has several sentences to ensure splitting. Here another sentence to make it longer.''')

len('Paragraph two: This is the second paragrah which') # 48
len('is quite long and will exceed our chunk size') # 44
len('limit for demenostration purpose.') # 33

33

**final**

In [ ]:
len('Paragraph one: This is the first paragraph.') # 43
len('It has two sentences.') # 21
len('Paragraph two: This is the second paragrah which') # 48
len('is quite long and will exceed our chunk size') # 44
len('limit for demenostration purpose.') # 33
len('Paragraph three: Short.') # 23

**chunk size after splitting should be less than or equal to provided chunk size**

In [153]:
from langchain.text_splitter import RecursiveCharacterTextSplitter

separators = ['\n\n','\n',' ','']

text="""Paragraph one: This is the first paragraph. It has two sentences.

Paragraph two: This is the second paragrah which is quite long and will exceed our chunk size limit
for demenostration purpose. It has several sentences to ensure splitting. Here another sentence to make it longer.

Paragraph three: Short."""

splitter = RecursiveCharacterTextSplitter(separators = separators,
                                 chunk_size = 50,
                                 chunk_overlap = 0                                 
                                )

chunks = splitter.split_text(text)
chunks

['Paragraph one: This is the first paragraph. It has',
 'two sentences.',
 'Paragraph two: This is the second paragrah which',
 'is quite long and will exceed our chunk size',
 'limit',
 'for demenostration purpose. It has several',
 'sentences to ensure splitting. Here another',
 'sentence to make it longer.',
 'Paragraph three: Short.']

In [155]:
for i,c in enumerate(chunks):
    print(f"Chunk {i} has {len(c)} characters")

Chunk 0 has 50 characters
Chunk 1 has 14 characters
Chunk 2 has 48 characters
Chunk 3 has 44 characters
Chunk 4 has 5 characters
Chunk 5 has 42 characters
Chunk 6 has 43 characters
Chunk 7 has 27 characters
Chunk 8 has 23 characters


In [157]:
from langchain.text_splitter import RecursiveCharacterTextSplitter

separators = ['\n\n','.',' ','']

text="""Paragraph one: This is the first paragraph. It has two sentences.

Paragraph two: This is the second paragrah which is quite long and will exceed our chunk size limit
for demenostration purpose. It has several sentences to ensure splitting. Here another sentence to make it longer.

Paragraph three: Short."""

splitter = RecursiveCharacterTextSplitter(separators = separators,
                                 chunk_size = 50,
                                 chunk_overlap = 0                                 
                                )

chunks = splitter.split_text(text)
chunks

['Paragraph one: This is the first paragraph',
 '. It has two sentences.',
 'Paragraph two: This is the second paragrah which',
 'is quite long and will exceed our chunk size',
 'limit\nfor demenostration purpose',
 '. It has several sentences to ensure splitting',
 '. Here another sentence to make it longer.',
 'Paragraph three: Short.']

In [159]:
for i,c in enumerate(chunks):
    print(f"Chunk {i} has {len(c)} characters")

Chunk 0 has 42 characters
Chunk 1 has 23 characters
Chunk 2 has 48 characters
Chunk 3 has 44 characters
Chunk 4 has 32 characters
Chunk 5 has 46 characters
Chunk 6 has 42 characters
Chunk 7 has 23 characters


### TokenTextSplitter

- If we want to chunks based on **tokens** then we use **TokenTextSplitter**

- 'Hello' is one token, but 'Hello' has 5 characters

- CharacterTextSplitter and RecursiveCharacterText splitter based on characters only

- But TokenTextSplitter completely based on token only

    - tokens can be words, special chars

**tiktoken**

- Official OpenAI tokenizer

- it converts text to tokens and tokens to text

- Count tokens in text

- Split texts so that it fits with token limits

- Estimate cost

- pip install tiktoken

In [7]:
import tiktoken

encoding = tiktoken.encoding_for_model('gpt-3.5-turbo')
encoding

<Encoding 'cl100k_base'>

In [9]:
text = 'I love learning Langchain and AI!'
tokens = encoding.encode(text)
tokens

[40, 3021, 6975, 23272, 8995, 323, 15592, 0]

In [11]:
len(tokens)

8

In [13]:
len(text)

33

**encode-decode**

- encode will convert a text to token id

- decode will convert token id to a text

In [16]:
encoding.encode('learning')

[21656]

In [22]:
encoding.encode('Langchain'),encoding.decode([27317])

([27317, 8995], 'Lang')

In [24]:
encoding.decode(tokens)

'I love learning Langchain and AI!'

In [30]:
word = 'I love learning Langchain and AI!'
t = encoding.encode(word)
for i in t:
    print(i, encoding.decode([i]))

40 I
3021  love
6975  learning
23272  Lang
8995 chain
323  and
15592  AI
0 !


In [32]:
text = 'I love learning Langchain and AI!'
words = text.split()
for i in words:
    print(i,encoding.encode(i))

I [40]
love [31153]
learning [21656]
Langchain [27317, 8995]
and [438]
AI! [15836, 0]


In [34]:
word = 'I love learning Langchain and AI!'
t = encoding.encode(word)
encoding.decode(t)

'I love learning Langchain and AI!'

In [36]:
encoding.n_vocab

100277

In [40]:
for i in range(50):
    print(encoding.decode([i]),end=' ')

! " # $ % & ' ( ) * + , - . / 0 1 2 3 4 5 6 7 8 9 : ; < = > ? @ A B C D E F G H I J K L M N O P Q R 

In [42]:
for i in range(2000,2050):
    print(encoding.decode([i]),end=' ')

for  — ten oin  ke ui  IN  table sub  Le  head  must //////////////// .util Context  order  mov over  contin  say static .Text  className pany  ter head rg  product This .”  But 70 loy  double sg  place .x message  information private  oper ced db "></ Param icle  week  prop table idget 

In [44]:
encoding.encode('Pranay')

[3617, 276, 352]

In [46]:
encoding.decode([3617, 276, 352])

'Pranay'

In [50]:
encoding.decode([3617]), encoding.decode([276]), encoding.decode([352])

('Pr', 'an', 'ay')

- But directly langchain has method

- **TokenTextSplitter**

- no need to use **tiktoken**

- **tiktoken** will give understanding of how words ==== tokens ==== token ids

- TokenTextSplitter by default will take **gpt-3.5-turbo** i.e **cl100k_base**

In [62]:
from langchain.text_splitter import TokenTextSplitter

text="""Paragraph one: This is the first paragraph. It has two sentences.
Paragraph two: This is the second paragraph which is quite long and will exceed our chunk size limit
for demenostration purpose. It has several sentences to ensure splitting. Here another sentence to make it longer.
Paragraph three: Short."""

splitter = TokenTextSplitter(chunk_size=15,
                            chunk_overlap=3)

# each chunk will have 15 tokens
# each token overlap of 3 tokens

chunks = splitter.split_text(text)
len(chunks)

6

In [64]:
chunks

['Paragraph one: This is the first paragraph. It has two sentences.',
 ' two sentences.\nParagraph two: This is the second paragraph which is',
 ' paragraph which is quite long and will exceed our chunk size limit\nfor dem',
 '\nfor demenostration purpose. It has several sentences to ensure splitting',
 ' to ensure splitting. Here another sentence to make it longer.\nParagraph',
 '\nParagraph three: Short.']

In [66]:
len(encoding.encode(chunks[0]))

14

In [68]:
chunks[2]

' paragraph which is quite long and will exceed our chunk size limit\nfor dem'

In [70]:
word = 'demonstration'
t = encoding.encode(word)
print(len(t))
for i in t:
    print(i,encoding.decode([i]))

3
67 d
7439 emon
55681 stration


In [72]:
encoding.encode(chunks[0])

[43265, 832, 25, 1115, 374, 279, 1176, 14646, 13, 1102, 706, 1403, 23719, 13]

In [74]:
encoding.encode('paragraph')

[28827]

In [81]:
# if we want to use different models

# the method is : from_tiktoken_encoder

dir(TokenTextSplitter)

['__abstractmethods__',
 '__annotations__',
 '__class__',
 '__delattr__',
 '__dict__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__getstate__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__le__',
 '__lt__',
 '__module__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__setattr__',
 '__sizeof__',
 '__slots__',
 '__str__',
 '__subclasshook__',
 '__weakref__',
 '_abc_impl',
 '_join_docs',
 '_merge_splits',
 'atransform_documents',
 'create_documents',
 'from_huggingface_tokenizer',
 'from_tiktoken_encoder',
 'split_documents',
 'split_text',
 'transform_documents']

In [87]:
from langchain.text_splitter import TokenTextSplitter

text="""Paragraph one: This is the first paragraph. It has two sentences.
Paragraph two: This is the second paragraph which is quite long and will exceed our chunk size limit
for demenostration purpose. It has several sentences to ensure splitting. Here another sentence to make it longer.
Paragraph three: Short."""

splitter = TokenTextSplitter.from_tiktoken_encoder(model_name='gpt-4',
                                                   chunk_size=15,
                                                   chunk_overlap=3)

# each chunk will have 15 tokens
# each token overlap of 3 tokens

chunks = splitter.split_text(text)
len(chunks)

5

In [89]:
chunks

['Paragraph one: This is the first paragraph. It has two sentences.\nParagraph',
 ' sentences.\nParagraph two: This is the second paragraph which is quite long and',
 ' quite long and will exceed our chunk size limit\nfor demenostration purpose',
 'enostration purpose. It has several sentences to ensure splitting. Here another sentence',
 ' Here another sentence to make it longer.\nParagraph three: Short.']